### PR TITLE
PS-7286: Add max allowed value check for encryption_key_id (8.0)

### PR DESCRIPTION
**Pull request by oleksandr-kachan**
There is no max allowed value check for encryption_key_id
while creating/altering table or tablespace. Because of this
the resulting value may contain any garbage in case provided value
exceeds underlying datatype limits.

Added check for max allowed value provided for encryption_key_id
in parser to be not bigger then UINT_MAX32-1. Added allowed values
check for innodb_default_encryption_key_id variable.

### CODE REVIEW

**Comment by oleksandr-kachan** [comment]
https://ps80.cd.percona.com/job/percona-server-8.0-param/793/

**Comment by satya-bodapati** [comment]
@oleksandr-kachan is this PR still valid after your changes? https://github.com/percona/percona-server/pull/3882/
If it is valid, please take over
If it is not valid, please bring those testcases to this PR

**Comment by oleksandr-kachan** [comment]
> @oleksandr-kachan is this PR still valid after your changes? #3882
> If it is valid, please take over
> If it is not valid, please bring those testcases to this PR

@satya-bodapati 
Updated this PR, will close  #3882

**Comment by oleksandr-kachan** [comment]
https://ps80.cd.percona.com/job/percona-server-8.0-param/794/

**Comment by satya-bodapati** [comment]
> > @oleksandr-kachan is this PR still valid after your changes? #3882
> > If it is valid, please take over
> > If it is not valid, please bring those testcases to this PR
> 
> @satya-bodapati
> Updated this PR, will close #3882

Did you take any parts of code from #3882?

**Comment by oleksandr-kachan** [comment]
> Did you take any parts of code from #3882?

@satya-bodapati 
Yes, I added check for min/max value for innodb_default_encryption_key_id variable but with my improvements. Also replaced my test with one from #3882.

File: mysql-test/suite/sys_vars/t/innodb_default_encryption_key_id_basic.test [hunk 1]
```diff
@@ -1,3 +1,62 @@
...
+--error ER_PARSE_ERROR
+CREATE TABLE t_below_0 (a varchar(255)) ENCRYPTION_KEY_ID=-1;
+
+CREATE TABLE t_MAX (a varchar(255)) ENCRYPTION_KEY_ID=4294967294;
+CREATE TABLE t_MIN (a varchar(255)) ENCRYPTION_KEY_ID=0;
+
+--error ER_ENCRYPTION_KEY_ID_VALUE_OUT_OF_RANGE
+ALTER TABLE t_MAX ENCRYPTION_KEY_ID=4294967295;
+--error ER_PARSE_ERROR
+ALTER TABLE t_MIN ENCRYPTION_KEY_ID=-1;
```
Comment (satya-bodapati) [inline mysql-test/suite/sys_vars/t/innodb_default_encryption_key_id_basic.test:94]: Why don't we get the same error code for both cases?

File: mysql-test/suite/sys_vars/r/innodb_default_encryption_key_id_basic.result [hunk 2]
```diff
@@ -1,3 +1,35 @@
 SELECT @@global.innodb_default_encryption_key_id;
 @@global.innodb_default_encryption_key_id
 0
+# restart:--early-plugin-load=keyring_file=keyring_file.so --keyring_file_data=MYSQL_TMP_DIR/keyring/keyring --default-table-encryption=ONLINE_TO_KEYRING --innodb-encryption-rotate-key-age=15 --innodb-encryption-threads=4 --innodb-default-encryption-key-id=4294967295 --log-error=MYSQL_TMP_DIR/my_restart.err
+# restart:--early-plugin-load=keyring_file=keyring_file.so --keyring_file_data=MYSQL_TMP_DIR/keyring/keyring --default-table-encryption=ONLINE_TO_KEYRING --innodb-encryption-rotate-key-age=15 --innodb-encryption-threads=4 --innodb-default-encryption-key-id=-1 --log-error=MYSQL_TMP_DIR/my_restart.err
+SET innodb_default_encryption_key_id=4294967295;
+Warnings:
+Warning	1292	Truncated incorrect innodb_encryption_key_id value: '4294967295'
```
Comment (satya-bodapati) [inline mysql-test/suite/sys_vars/r/innodb_default_encryption_key_id_basic.result:8]: after every SET, please show the current value and global SELECT @@session. and @@global.

File: mysql-test/suite/sys_vars/t/innodb_default_encryption_key_id_basic.test [hunk 3]
```diff
@@ -1,3 +1,62 @@
...
+--let $restart_parameters=restart:--early-plugin-load="keyring_file=$KEYRING_PLUGIN" --keyring_file_data=$MYSQL_TMP_DIR/keyring/keyring --default-table-encryption=ONLINE_TO_KEYRING --innodb-encryption-rotate-key-age=15 --innodb-encryption-threads=4 --innodb-default-encryption-key-id=4294967295 --log-error=$error_log
+--replace_result $MYSQL_TMP_DIR MYSQL_TMP_DIR
+--source include/restart_mysqld.inc
+
+--let ABORT_ON=NOT_FOUND
+--let SEARCH_FILE=$error_log
+--let SEARCH_PATTERN=\[Server\] option 'innodb-default-encryption-key-id': unsigned value 4294967295 adjusted to 4294967294\.
+--source include/search_pattern_in_file.inc
+
+--let $restart_parameters=restart:--early-plugin-load="keyring_file=$KEYRING_PLUGIN" --keyring_file_data=$MYSQL_TMP_DIR/keyring/keyring --default-table-encryption=ONLINE_TO_KEYRING --innodb-encryption-rotate-key-age=15 --innodb-encryption-threads=4 --innodb-default-encryption-key-id=-1 --log-error=$error_log
```
Comment (satya-bodapati) [inline mysql-test/suite/sys_vars/t/innodb_default_encryption_key_id_basic.test:15]: where the keyring_file_data deleted?

File: mysql-test/suite/sys_vars/t/innodb_default_encryption_key_id_basic.test [hunk 4]
```diff
@@ -1,3 +1,62 @@
---skip Not implemented
-
 SELECT @@global.innodb_default_encryption_key_id;
+
+--let $error_log=$MYSQL_TMP_DIR/my_restart.err
```
Comment (satya-bodapati) [inline mysql-test/suite/sys_vars/t/innodb_default_encryption_key_id_basic.test:3]: This should be also deleted? (or is it already deleted somewhere below?)

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: mysql-test/suite/sys_vars/t/innodb_default_encryption_key_id_basic.test (hunk 1 above)
Comment (oleksandr-kachan) [reply mysql-test/suite/sys_vars/t/innodb_default_encryption_key_id_basic.test:94]: @satya-bodapati, min and max value checks are done in different places in this case. For max value we directly check in parser rules which we can change and use whichever error we want. The min value check is done by parser itself based on datatype assigned to the symbol which is unsigned in this case - "ENCRYPTION_KEY_ID_SYM opt_equal real_ulong_num". We could in general change datatype for ENCRYPTION_KEY_ID_SYM to signed and do check its min value manually. Do you think we should follow this way?

Comment (satya-bodapati) [reply mysql-test/suite/sys_vars/t/innodb_default_encryption_key_id_basic.test:94]: Got it! thanks

File: share/messages_to_clients.txt [hunk 5]
```diff
@@ -9480,6 +9480,9 @@ ER_SYSTEM_KEY_ROTATION_CANT_GENERATE_NEW_VERSION
 ER_DTE_ENCRYPTION_THREADS_ACTIVE
   eng "In order to change default_table_encryption, disable encryption threads. Please set number of innodb_encryption_threads to 0."
 
+ER_ENCRYPTION_KEY_ID_VALUE_OUT_OF_RANGE
+  eng "The requested value for the encryption_key_id exceeds the maximum allowed %lu."
```
Comment (satya-bodapati) [inline share/messages_to_clients.txt]:
```suggestion
ER_ENCRYPTION_KEY_ID_VALUE_OUT_OF_RANGE
  eng "The requested value %lu for the encryption_key_id exceeds the maximum allowed 4294967295"
```

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: share/messages_to_clients.txt (hunk 5 above)
Comment (oleksandr-kachan) [reply share/messages_to_clients.txt]: Done

File: mysql-test/suite/sys_vars/t/innodb_default_encryption_key_id_basic.test [hunk 6]
```diff
@@ -1,3 +1,108 @@
...
 SELECT @@global.innodb_default_encryption_key_id;
+
+SET innodb_default_encryption_key_id=0;
+SELECT @@session.innodb_default_encryption_key_id;
+SELECT @@global.innodb_default_encryption_key_id;
+
+SET global innodb_default_encryption_key_id=0;
+SELECT @@session.innodb_default_encryption_key_id;
+SELECT @@global.innodb_default_encryption_key_id;
+
```
Comment (satya-bodapati) [inline mysql-test/suite/sys_vars/t/innodb_default_encryption_key_id_basic.test:63]: can you set to non-zero before SET PERSIST? that way we will know whether SET PERSIST changed the value or not?

Comment (satya-bodapati) [reply mysql-test/suite/sys_vars/t/innodb_default_encryption_key_id_basic.test:63]: Same for PERSIST_ONLY

**Review by satya-bodapati — CHANGES_REQUESTED** [review]
Looks good mostly. One last minor request

File: mysql-test/suite/sys_vars/r/innodb_default_encryption_key_id_basic.result [hunk 7]
```diff
@@ -1,3 +1,140 @@
...
+SELECT @@global.innodb_default_encryption_key_id;
+@@global.innodb_default_encryption_key_id
+1
+SET persist innodb_default_encryption_key_id=0;
+SELECT @@session.innodb_default_encryption_key_id;
+@@session.innodb_default_encryption_key_id
+1
+SELECT @@global.innodb_default_encryption_key_id;
+@@global.innodb_default_encryption_key_id
+0
```
Comment (satya-bodapati) [inline mysql-test/suite/sys_vars/r/innodb_default_encryption_key_id_basic.result:102]: is this expected SET PERSIST behaviour? This changes the global value and not the session value. Can you please check SET PERSIST with another session variable?

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: mysql-test/suite/sys_vars/r/innodb_default_encryption_key_id_basic.result (hunk 7 above)
Comment (oleksandr-kachan) [reply mysql-test/suite/sys_vars/r/innodb_default_encryption_key_id_basic.result:102]: @satya-bodapati, to my knowledge this is expected - SET PERSIST updates global variable only.

Comment (satya-bodapati) [reply mysql-test/suite/sys_vars/r/innodb_default_encryption_key_id_basic.result:102]: ok. Can you also check the Jenkins results. the test is skipped on release builds
201002 11:06:48 [ 77%] sys_vars.innodb_default_encryption_key_id_basic w4  [ skipped ]  Not implemented

Please resubmit to Jenkins (some issue with debug builds)

**Comment by oleksandr-kachan** [comment]
https://ps80.cd.percona.com/job/percona-server-8.0-param/795/

**Review by satya-bodapati — APPROVED** [review]

**Comment by satya-bodapati** [comment]
Analyzed Jenkins results. Looks good

**Pull request merged**